### PR TITLE
[Relay] improve SimplifyClipAndConsecutiveCast pass

### PR DESCRIPTION
**Pull request by kfeng123**
This PR improves SimplifyClipAndConsecutiveCast pass in several aspects.
First, it allows that the dtype of `clip` and the dtype of the last `cast` differ.
Second, it allows that there are arbitrary number of `cast` after `clip`. This is achieved via the recursion of patterns.
Both of the above two cases araise from the compilation of a prequantized model.

See the [pre-RFC](https://discuss.tvm.apache.org/t/on-the-applications-of-the-composition-of-dataflow-patterns/15344?u=kfeng123)

### CODE REVIEW

**Comment by kfeng123** [comment]
Perhaps the failure of test is due to the optimization of this PR: this PR may reduce the memory the model requires.
I file a [issue](https://github.com/apache/tvm/issues/15365).
Hopefully, some experts of usmp may help to verify this issue.

**Comment by kfeng123** [comment]
> Perhaps the failure of test is due to the optimization of this PR: this PR may reduce the memory the model requires. I file a [issue](https://github.com/apache/tvm/issues/15365). Hopefully, some experts of usmp may help to verify this issue.

For now I just modify the ground truth of the test case to pass tests. Not sure if this is correct. Correct me if the original ground truth is indeed correct.

File: src/relay/transforms/simplify_expr.cc [hunk 1]
```diff
@@ -160,7 +160,7 @@ class SimplifyConsecutiveCast : public DFPatternRewrite {
   DFPattern cast1_;
 };
 
-bool CheckDataTypeMaxMinValue(DataType dtype, double min_value, double max_value) {
+bool CheckDataTypeMaxMinValue(DataType dtype, double min_value, double max_value, int mode = 0) {
```
Comment (Hzfengsy) [inline src/relay/transforms/simplify_expr.cc:163]: Please comment the meaning of different `mode`

Comment (kfeng123) [reply src/relay/transforms/simplify_expr.cc:166]: Done

**Review by Hzfengsy — APPROVED** [review]

**Pull request merged**